### PR TITLE
CP-17506: Add new SM capability Vdi_mirror and update VDI and SR operations.

### PR DESCRIPTION
**Pull request by sharady**

### CODE REVIEW

File: ocaml/xapi/xapi_vdi.ml [hunk 1]
```diff
@@ -44,8 +45,9 @@ let check_operation_error ~__context ?(sr_records=[]) ?(pbd_records=[]) ?(vbd_re
 	      has a current_operation itself
 	   5. HA prevents you from deleting statefiles or metadata volumes
 	   *)
-	if List.exists (fun (_, op) -> op <> `copy) current_ops
-	then Some(Api_errors.other_operation_in_progress,["VDI"; _ref])
+	if (List.exists (fun (_, op) -> op <> `copy) current_ops) &
+		not ((op = `destroy) & (List.exists (fun (_, op) -> op = `mirror) current_ops))
```
Comment (simonjbeaumont) [inline ocaml/xapi/xapi_vdi.ml]: I think we should break this clause out into a separate check for clarity.

Comment (simonjbeaumont) [inline ocaml/xapi/xapi_vdi.ml]: Also, the infix `&` operator is deprecated for a while now; should be `&&`:

```ocaml
$ ocaml
        OCaml version 4.02.1

# (&);;
Characters 0-3:
  (&);;
  ^^^
Warning 3: deprecated: Pervasives.&
Use (&&) instead.
- : bool -> bool -> bool = <fun>
```

**Comment by simonjbeaumont** [comment]
Your commit message needs changing 'CP-17506: Add new capability Vdi_mirror for NutanixSR' to remove branding. Also it's the opposite of what we're trying to achieve with this change :) There's nothing specific to any SR in your pull-request and we'll be adding this capability to many existing SM plugins.

**Comment by simonjbeaumont** [comment]
I'll add the needs updating label because there are some minor fixups required. However, @jonludlam is going to give an additional review on this to check I've not missed anything.

**Comment by jonludlam** [comment]
Looks good to me, once the bits @simonjbeaumont has pointed out have been done.

**Comment by simonjbeaumont** [comment]
Thanks @jonludlam. @sharady: Jon pointed out that we can't merge this until the existing storage drivers are advertising this capability. Have you created the ticket for them? Also, he suggests we run this change through the SXM XenRT suite before merging which I think is a good idea.

**Comment by sharady** [comment]
Unblocking this PR, As SM changes has gone into `trunk-ring3/sm.pg.git` as part of CP-17867.

**Pull request merged**